### ISSUE TITLE
Borg 1.1 running for days

### ISSUE THREAD

**Issue by EthraZa** [borgbackup/borg#3299]
I'm trying the first run of Borg (1.1.1) on an old Linux server (Ubuntu 10.04) with millions of small XML files. It's now running since saturday night. (Repository is in another server over 1GB LAN)
The fact is it is taking longer than expected and is disturbing the server, since Borg is using up to 23% of the 16GB RAM.
By reading the docs I can't determine if there is some flag that could improve performance with lots of small files on EXT4 FS. Like, is nobsdflags for me?

I'm going to kill its process now... hope that kill -15 on the borg subproccess is the appropriate way.

By the way, I'm using the following command to see what Borg is doing and it may help other people if added to the manual:
`watch -n0,1 "lsof -c borg -l -X|grep -E 'r   REG|w   REG'"`

**Comment by ThomasWaldmann** [comment]
Lots of small files are never going to be a super-fast thing (that's not specific to borg, rather a general thing).

Memory needs of borg depend on some parameters (including files count, total chunks count), there's a formula in the docs. You could switch off the files cache to save some memory, but otoh, the files cache is what makes borg so super fast for 2nd+ backups, so that is no easy choice.

`--nobdsflags` will make it a bit faster on Linux as it will save 1 open() call per file. This is especially relevant for 2nd+ backups *with* files cache as then it will not open unchanged files at all.

BTW, borg does checkpoints, so killing it decently (SIGTERM, SIGINT) will leave a checkpoint archive in the repo. So if you just start the same backup later again when the server is less busy, it will save a lot of time due to deduplication with the checkpoint archive.

1.1.2 is out meanwhile with some more fixes.

**Comment by ThomasWaldmann** [comment]
Are these millions of XML files a static collection or are they changing?

**Comment by EthraZa** [comment]
The XMLs never change but we have thousands more arriving every day in the same directories.

**Comment by ThomasWaldmann** [comment]
If you can't afford the memory overhead the XML files are causing, an option could be to exclude the directory from borg backup and just make a uncompressed tar archive of them all before borg runs (and backup the tar with borg). That way, you will have way less objects in the repository, in the caches and indexes (assuming you start from scratch with a new empty repo). How deduplication behaves for the tar depends a bit: if there are longer (>= 2MiB) identical sequences in the tar, it might work ok.

If it is just the run time that is bothering you, you should just try to complete the initial backup somehow (even if needs multiple partial runs). The 2nd+ backup will be faster (with --nobsdflags).

prune will not kill the most recent checkpoint that is not yet superceded by a successful backup.

**Comment by EthraZa** [comment]
Today I decided to run borg with --progress and --list to see what it is doing...
I'm seeing status "M" for lots of files, files that do not and cannot change, for ages!

Example:
`-rw-r--r-- 1 web1 client0 917 2010-11-30 20:18 /var/www/.../debug.css`

Is that right?

**Comment by ThomasWaldmann** [comment]
Well, if borg is seeing these files for the first time, status should be "A" like added. If it sees the same file again, it should be "U" like unchanged. "M" is file that was somehow modified.

Modified means that borg detected a change relative to the values in the files cache, which are:
- ctime, size, inode number (1.1 default)
- mtime, size, inode number (1.0 default, option for 1.1)
- some other combinations of these

If you're seeing M again and again and the file content did not change, it could be:
- inode numbers not being stable on the filesystem you use (e.g. network fs, some fuse fs)
- the file's inode being modified, e.g. chown, chmod, ln, ... - this will trigger a ctime change
- the file's mtime being touched

**Comment by EthraZa** [comment]
Gosh, with "M", this 1st run will never end!

Using that same example file:
```
# stat /var/www/.../debug.css
  File: "/var/www/.../debug.css"
  Size: 917             Blocks: 8          IO Block: 4096   arquivo comum
Device: fb00h/64256d    Inode: 66847710    Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 5004/    web1)   Gid: ( 5036/ client0)
Access: 2017-11-08 20:47:28.167838395 -0200
Modify: 2010-11-30 20:18:14.000000000 -0200
Change: 2015-09-25 17:30:48.435829946 -0300
```
So, based in what you said, it must be the inode, and it beeing an ext4 fs, I can't even imagine what could be touching these files inodes!

This server has ISPConfig, so it could be doing something. But I don't imagine it chmoding or chowning millions of files (the XMLs) everyday, I guess these operations would take lots of time to execute in this case, no?

I don't think there is a way to know if the inode get chaged! You sad it would be change right? Or is just some flag that is changing? If so, there is something I can look for to investigate it?

**Comment by ThomasWaldmann** [comment]
you can show the inode number with `ls -i` or with stat. if a file repeatedly shows M or A in borg runs, check that + ctime/mtime and size. ext4 shouldn't have inode instability issues.

"auto,lz4" is almost the same as "lz4", btw. auto mode is usually used with a more expensive compression like zlib or lzma (and uses lz4 internally as compressibility predictor).

**Comment by EthraZa** [comment]
I guess I got Auto now. The manual didn't let it clear if just lz4 would not recompress already compressed files.

The stat for that example file not changed betwwen yestarday and today... Now I'll setup a test repo only on that directory to see how it goes day to day. 
For now, I was forced to stop trying to use Borg for the /var/www and keep backuping it with rsync, but I hope we can determine the cause of this MMs poping up in the backup. :)

I immensely appreciate your support

**Comment by EthraZa** [comment]
Well I just setup my test repo to backup this folder with 1.1GB of very old JS and CSS files that never changes.
I run it 4 times in a row now. The 1st all files get A. The other 3 times, the same 3 files always get A.

On janeway:
```
# ssh root@picard "borg create --progress --list --filter AME --stats --show-rc --compression auto,lz4 --exclude-caches --nobsdflags --exclude-from exclude.txt root@janeway:/media/borg/repo/root@picard_TST::'{hostname}-{now}' /var/www/clients/client0/web1/web/libweb/" 2>&1|tee tst.log
A /var/www/clients/client0/web1/web/libweb/ext-core-3.4.0/ext-core.jsext-core.js
A /var/www/clients/client0/web1/web/libweb/ext-core-3.4.0/ext-core.jsb2
A /var/www/clients/client0/web1/web/libweb/ext-core-3.4.0/ext-core-debug.js
------------------------------------------------------------------------------  
Archive name: picard.ghsix.com.br-2017-11-09T13:11:52
Archive fingerprint: 85f495dd144a7ccd8a42aad3eec71b32ff7aaf8266b1c3b6c87db8cddffddc7b
Time (start): Thu, 2017-11-09 13:11:54
Time (end):   Thu, 2017-11-09 13:12:20
Duration: 25.89 seconds
Number of files: 45009
Utilization of max. archive size: 0%
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
This archive:              986.55 MB            324.51 MB              3.92 kB
All archives:                3.95 GB              1.30 GB            255.29 MB

                       Unique chunks         Total chunks
Chunk index:                   15527               180800
------------------------------------------------------------------------------
```
On picard:
```
# stat /var/www/clients/client0/web1/web/libweb/ext-core-3.4.0/ext-core-debug.js
  File: "/var/www/clients/client0/web1/web/libweb/ext-core-3.4.0/ext-core-debug.js"
  Size: 401191          Blocks: 784        IO Block: 4096   arquivo comum
Device: fb00h/64256d    Inode: 18226886    Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 5004/    web1)   Gid: ( 5036/ client0)
Access: 2017-10-15 01:18:18.817738340 -0200
Modify: 2011-11-18 14:06:22.000000000 -0200
Change: 2015-09-25 17:30:53.234772169 -0300
```

Just that already looks strange to me, but tomorrow I will run it again to see how it behaves.

**Comment by ThomasWaldmann** [comment]
@EthraZa when you save "run it 4 times in a row", it means that it successfully ran and completed with rc=0?

Or did it crash or get interrupted in the middle somehow?

And these 4 runs were all borg 1.1.2?

**Comment by ThomasWaldmann** [comment]
BTW, see also the faq entry why the newest file(s) will always show up as "A".

**Comment by EthraZa** [comment]
Hi, yes, 4 times in a row with rc 0, and yes, Borg 1.1.2 on both servers.

**Comment by EthraZa** [comment]
Also, in the manual: Linux: glibc >= 2.13 (ok for most supported Linux releases). Older glibc releases are untested and may not work.

On this old servers (both are):
```
# ldd --version
ldd (Ubuntu EGLIBC 2.11.1-0ubuntu7.21) 2.11.1
Copyright (C) 2009 Free Software Foundation, Inc.
```
It could be the reason for something like this?

**Comment by ThomasWaldmann** [comment]
about glibc: not sure. There are some calls missing in older glibcs, but I'ld rather expect it to crash then (especially when you use the binary). Guess that ubuntu is out of security support anyway, isn't it?

**Comment by EthraZa** [comment]
WOW! To test the manual suggestion, I created a new file inside that path so it would be the newst one. 
The file is there, I can vi or cat it, I can even access it over http, it is there... But Borg is not adding it to the backup. It's invisible to Borg!

**Comment by EthraZa** [comment]
Yep, creating the new file (one that os not part of ignored files), did the trick about the "A"s. Now we need to wait tomorrow to test again and see the inodes.

**Comment by ThomasWaldmann** [comment]
It is adding the newest file to the backup, just not to the cache, see the faq for an explanation.
If that solves your "it's all A" problem: are all your other files having the same ctime or mtime?

**Comment by EthraZa** [comment]
Yep, the old files are very similar in their times.
```
# stat ext-core-3.4.0/ext-core.js
  File: "ext-core-3.4.0/ext-core.js"
  Size: 91868           Blocks: 184        IO Block: 4096   arquivo comum
Device: fb00h/64256d    Inode: 18226430    Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 5004/    web1)   Gid: ( 5036/ client0)
Access: 2017-10-16 22:00:08.877736029 -0200
Modify: 2011-11-18 14:06:18.000000000 -0200
Change: 2015-09-25 17:30:53.234772169 -0300

# stat ext-4.2.3.1477/ext-all.js 
  File: "ext-4.2.3.1477/ext-all.js"
  Size: 1593212         Blocks: 3112       IO Block: 4096   arquivo comum
Device: fb00h/64256d    Inode: 66259337    Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 5004/    web1)   Gid: ( 5036/ client0)
Access: 2017-10-14 13:23:06.647524955 -0300
Modify: 2014-09-02 11:24:08.000000000 -0300
Change: 2015-09-25 17:30:52.604811264 -0300
```
But the "A"s kind of makes sense and are explained in the FAQ, my real concern are the "M"s, they are preventing me to use Borg. They are not happen in this new repo yet, let's see tomorrow.

**Comment by EthraZa** [comment]
New day, new doubts.
I did the test today with the subset of 1GB of my backup, and Borg didn't "M"s the files so, it looks like it's nothing wrong with the FS or disk.

**Considerations:**
I don't know a way to count how many XMLs I have now, what I can tell is how many new XMLs come per day, and we are receiving about 200.000 new XMLs a day. On the Christmas period this number will almost double, so between November's end and December it will be about 380.000 new files/day. These system run since 2010, of course the numbers coming grewing by the years, but we can do a close call math now. About half of these XMLs, are .xml.gz (newer different system), the other half are just .xml (older system).

**Now the big question:**
Maybe this amount of files are "overflowing" the Borg cache? I mean, if the cache has any limit, some files already backuped may not be in the cache anymore on a 2nd+ run, so it would be chunked again "M".
If so, what do I need to do to up this cache limit? 
Or maybe, how to make a different cache for some XML directories so Borg can keep up with these millions files? Just make a different call to borg create will do that, or I need a different repository or a different user?

**Comment by ThomasWaldmann** [comment]
There is no limit in file count, but there is a limit in generations (== backup runs) for which the cache remembers stuff, see the environment vars docs.

If a file entry in the cache is (intentionally) forgotten after some backups in which it was not seen, the file will be "A" if it appears again in a backup.

**Comment by EthraZa** [comment]
In [the manual](http://borgbackup.readthedocs.io/en/stable/faq.html#it-always-chunks-all-my-files-even-unchanged-ones) is says:
> if you do daily backups of 26 different data sets A, B, C, ..., Z
> ...the files from A will be already forgotten when you repeat the same backups on the next day 

Data sets are 
`Borg create ... /d1 /d2`
or
`Borg create ... /d1`
`Borg create ... /d2`
?
My actual backup from picard is:
`borg create ... /etc /home/scripts /usr/local/ispconfig /var/spool/cron /var/lib/automysqlbackup/daily /var/log/mysql /var/www`
Whre the problem seems to be lying in the size of /var/www, if I take it out, it goes way fast in the backup.

By the way, there is some way to know why Borg is given "M" to a file? 
I mean, can we know if it is mtime, size, name, inode, missing chunk, missing in cache or what?

**Comment by ThomasWaldmann** [comment]
each borg create invocation has a set of input files, that's what's meant with data set.

**Comment by EthraZa** [comment]
Last friday I found out prune has eating my checkpoint full of XMLs, so I had to start it over, now I'll rename it and prune with prefix so it doen't happen again. 
I've let a borg running from sturday at 7am till monday at 2am and then I had to kill it. I will try to continue it next night. For now I have added the XML paths to exclude so borg can keep backuping the rest.

First I tough it was normal, maybe it is, not sure, but I'm running borg from server Janeway to Picard backuping it back to Janeway, and I'm noticing that borg shows up three times on the ps of both servers, but when I run borg on my desktop, it shows only once.
For example...

on janeway:
```
# ps -ef|grep borg
root     12399 17314  0 02:17 ?        00:00:00 ssh root@picard borg prune --prefix '{hostname}-' --list --show-rc --keep-daily 7 --keep-weekly 4 --keep-monthly 6 --keep-yearly 1 root@janeway:/media/borg/repo/root@picard
root     12483 12406  0 02:17 ?        00:00:00 borg serve --umask=077
root     12484 12483 44 02:17 ?        00:06:40 borg serve --umask=077
```
on picard:
```
# ps -ef|grep -i borg
root      8507  4843  0 02:31 pts/0    00:00:00 grep --color -i borg
root     21501 21199  0 02:17 ?        00:00:00 borg prune --prefix {hostname}- --list --show-rc --keep-daily 7 --keep-weekly 4 --keep-monthly 6 --keep-yearly 1 root@janeway:/media/borg/repo/root@picard
root     21521 21501 23 02:17 ?        00:03:23 borg prune --prefix {hostname}- --list --show-rc --keep-daily 7 --keep-weekly 4 --keep-monthly 6 --keep-yearly 1 root@janeway:/media/borg/repo/root@picard
root     21564 21521  0 02:17 ?        00:00:08 ssh root@janeway borg serve --umask=07
```
Is that right when we are running borg over ssh, or can that be the root of my headaches?

**Comment by EthraZa** [comment]
Thanks to Borg being so hard on memory usage and knocking down performance of this old server, I have managed to schedule the transport of the most heavy DBs out of this server to a new one.

What I have now for this last days is Borg running at night time, and taking exactly 4 hours to complete, just adding (A) about ten files every day.
I let the millions of XML files to Rsync for now. It is tooking about 10 hours every day, but since it is does not use much RAM or CPU we can afford let it running. And by limiting its bandwitch, we are limiting its IO as well, so it is slow but nice to the server.

**The Borg usage:**
What I noticed of Borg running this days is it uses about 7.2GB of RAM just to load up its indexes. And while watching it run, I saw it topings 7.5GB of RAM in use. When the server does not have this RAM available, the swap kicks in and there lies the problem.

Now, I'm not sure about that, but I'm guessing by the last times I tryed, that when I have as low as 2GB or 4GB of RAM available, and Borg wants 7.5GB to work, maybe something is going wrong with the index access and Borg starts to think all files are modified (M) and that leads to more delays than a server extra loaded with swaping.

**Comment by ThomasWaldmann** [comment]
New server: ok, so you see ~10 A(dded) files - how many M(odified) do you see? Should be only the really modified files, not all.

Yes, borg can assimilate substantial amounts of RAM, depends on files count, chunks count in repo.

When it needs swap to satisfy the memory needs, it of course will get very slow, but it should not show different function. Except if the memory needs can not be satisfied, then it'll crash.

So, I'ld guess you maybe had switched off the files cache to save memory when you saw "all M". Or you had switched from borg 1.0 to 1.1 (which changes the default mode for the files cache).

BTW, today I read that restic needs even more RAM than borg. https://blog.stickleback.dk/borg-or-restic/

**Comment by EthraZa** [comment]
By no means I'm complaining about RAM usage, I totally understand why Borg needs it.
I've started on Borg 1.1.1 and now I'm on 1.1.2.
I never disabled the cache, the switchs are still [the same](https://github.com/borgbackup/borg/issues/3299#issuecomment-343024051).

I just noticed it, when I added the XML dirs it started to M all files, at least while I was watching it, like the css example above. I'm not sure why, just guessing that it could be possible it growed so much in RAM usage that it leaked somewhere.
The truth is I don't know how to check it, I just think it may be something to be investigated by someone more qualified then me.

For now it may be better to close this issue, since it's getting long on supositions that I can not prove, and will not be able to test more for the coming weeks or so while I'll be migrating some servers.

Thank you for you patience and help.
